### PR TITLE
scripts: parametrize systemvm, router restart

### LINKED ISSUE #6220: Parametrize data source table for virtual routetrs to support migration
In a enterprise migration scenario, you could face a situation to restart hundreds of routers which may cause issues with operation of enterprise service like plannable downtime window, even if you split restart by zones.

The included scripted tool _setup/bindir/cloud-sysvmadm_ sources routers list from the [table cloud.vm_instance](https://github.com/apache/cloudstack/blob/133b195688874ad63a947ae6b3fd369b671f2597/setup/bindir/cloud-sysvmadm.in#L82).

Possibly, there are many possible approaches how to establish partial batches if there are many routers.

For example:

* Source alternatively from an external text file instead of direct DB connection
* Parametrize table name allowing cloning subsets to temporary data tables in the DB
* Provide a short list of router ID/names to restart individually

Here are some code lines how we implemented a parameter for an alternate table name, and it worked for us.

```
table="vm_instance"
..
while getopts 'sarhnvd:m:u:p:t:T:l:z:' OPTION
  T)	table="$OPTARG"
  		;;
  l)	LOGFILE="$OPTARG"
 ..

stop_start_system() {
secondary=(`mysql -h $db --user=$user --password=$password --skip-column-names -U cloud -e "select id from $table ...

```

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #6220

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

```
[root@ref-trl-3339-v-M7-abhishek-kumar-mgmt1 ~]# /usr/bin/cloudstack-sysvmadm

The tool stopping/starting running system vms and domain routers 

Usage: cloudstack-sysvmadm: [-d] [-u] [-p] [-m] [-s] [-r] [-a] [-t] [-n] [-z] [-i] [-j] [-v] [-i] [-j]

 -d - cloud DB server ip address, defaulted to localhost if not specified 
 -u - user name to access cloud DB, defaulted to root if not specified 
 -p - cloud DB user password, defaulted to no password if not specified 

 -m - the ip address of management server, defaulted to localhost if not specified

 -s - stop then start all running SSVMs and Console Proxies 
 -r - stop then start all running Virtual Routers
 -a - stop then start all running SSVMs, Console Proxies, and Virtual Routers 
 -n - restart all Guest networks 
 -t - number of parallel threads used for stopping Domain Routers. Default is 10.
 -l - log file location. Default is cloud.log under current directory.
 -z - do restart only for the instances in the specific zone. If not specified, restart will apply to instances in all zones
 -v - do restart all VPCs in the entire system
 -i - do restart on instances with specific IDs. Comma-separated instance IDs to be provided. Works ony when restarting SSVMs, CPVMs or VRs
 -j - do restart on instances from the specified database table. If not specified, restart will apply on instances present in cloud.vm_instance database table. Works ony when restarting SSVMs, CPVMs or VRs
[root@ref-trl-3339-v-M7-abhishek-kumar-mgmt1 ~]# /usr/bin/cloudstack-sysvmadm -a -u root -p P@ssword123 -i 9b802cd9-8d97-42dc-9610-2650a2831fee,2f1d6bc1-7143-42b7-a5b7-2383d4eddcb9

Stopping and starting 1 secondary storage vm(s) with id=9b802cd9-8d97-42dc-9610-2650a2831fee,2f1d6bc1-7143-42b7-a5b7-2383d4eddcb9...
Done stopping and starting secondary storage vm(s) with id=9b802cd9-8d97-42dc-9610-2650a2831fee,2f1d6bc1-7143-42b7-a5b7-2383d4eddcb9

Stopping and starting 1 console proxy vm(s) with id=9b802cd9-8d97-42dc-9610-2650a2831fee,2f1d6bc1-7143-42b7-a5b7-2383d4eddcb9...
Done stopping and starting console proxy vm(s)  with id=9b802cd9-8d97-42dc-9610-2650a2831fee,2f1d6bc1-7143-42b7-a5b7-2383d4eddcb9.

Stopping and starting 0 running routing vm(s) with id=9b802cd9-8d97-42dc-9610-2650a2831fee,2f1d6bc1-7143-42b7-a5b7-2383d4eddcb9... 
```

### CODE REVIEW

**Review by rohityadavcloud — APPROVED** [review]
needs manual qa, otherwise LGTM

File: setup/bindir/cloud-sysvmadm.in [hunk 1]
```diff
@@ -72,21 +76,36 @@ do
...
+        ;;
+  j)    vmtable="$OPTARG"
   esac
 done
 
 
-
+prapare_ids_clause() {
+  if [[ ! -z "$vmidsclause" ]]; then
+    return
```
Comment (weizhouapache) [inline setup/bindir/cloud-sysvmadm.in:90]: it looks line 89 to line 91 are not required.

ideally to have option to upgrade VR with livepatch.

Comment (shwstppr) [reply setup/bindir/cloud-sysvmadm.in:90]: @weizhouapache since same method is called twice, added this check to not generate vmidsclause twice

Comment (weizhouapache) [reply setup/bindir/cloud-sysvmadm.in:90]: ok, makes sense @shwstppr

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: setup/bindir/cloud-sysvmadm.in [hunk 2]
```diff
@@ -23,7 +23,7 @@
...
...ted to localhost if not specified \n -u - user name to access cloud DB, defaulted to "root" if not specified \n -p - cloud DB user password, defaulted to no password if not specified \n\n -m - the ip address of management server, defaulted to localhost if not specified\n\n -s - stop then start all running SSVMs and Console Proxies \n -r - stop then start all running Virtual Routers\n -a - stop then start all running SSVMs, Console Proxies, and Virtual Routers \n -n - restart all Guest networks \n -t - number of parallel threads used for stopping Domain Routers. Default is 10.\n -l - log file location. Default is cloud.log under current directory.\n -z - do restart only for the instances in the specific zone. If not specified, restart will apply to instances in all zones\n -v - do restart all VPCs in the entire system\n -i - do restart on instances with specific IDs. Comma-separated instance IDs to be provided. Works ony when restarting SSVMs, CPVMs or VRs\n -j - do restart on instances from the specified database table. If not specified, restart will apply on instances present in cloud.vm_instance database table. Works ony when restarting SSVMs, CPVMs or VRs\n\n" $(basename $0) >&2
```
Comment (weizhouapache) [inline setup/bindir/cloud-sysvmadm.in:26]: duplicated `[-i] [-j]`

Comment (weizhouapache) [reply setup/bindir/cloud-sysvmadm.in:26]: @shwstppr 
can you organize this by multiple lines ? it is hard to review.

**Review by weizhouapache — APPROVED** [review]
code lgtm
left 2 minor comments

File: setup/bindir/cloud-sysvmadm.in (hunk 2 above)
Comment (shwstppr) [reply setup/bindir/cloud-sysvmadm.in]: @weizhouapache refactored

File: setup/bindir/cloud-sysvmadm.in [hunk 3]
```diff
@@ -23,7 +23,26 @@
...
+  usage_content="
+  The tool for stopping/starting running system vms and domain routers
+
+  Usage: %s: [-d] [-u] [-p] [-m] [-s] [-r] [-a] [-t] [-n] [-z] [-i] [-j] [-v] [-i] [-j]
```
Comment (weizhouapache) [inline setup/bindir/cloud-sysvmadm.in:29]: double ` [-i] [-j] `

File: setup/bindir/cloud-sysvmadm.in [hunk 4]
```diff
@@ -23,7 +23,26 @@
...
+   -m - the ip address of management server, defaulted to localhost if not specified
+   -s - stop then start all running SSVMs and Console Proxies
+   -r - stop then start all running Virtual Routers
+   -a - stop then start all running SSVMs, Console Proxies, and Virtual Routers
+   -n - restart all Guest networks
+   -t - number of parallel threads used for stopping Domain Routers. Default is 10
+   -l - log file location. Default is cloud.log under current directory
+   -z - do restart only for the instances in the specific zone. If not specified, restart will apply to instances in all zones
+   -v - do restart all VPCs in the entire system
+   -i - do restart on instances with specific IDs. Comma-separated instance IDs to be provided. Works ony when restarting SSVMs, CPVMs or VRs
```
Comment (weizhouapache) [inline setup/bindir/cloud-sysvmadm.in:43]: ony -> only
same in line 43 and line 44

File: setup/bindir/cloud-sysvmadm.in [hunk 5]
```diff
@@ -23,7 +23,26 @@
...
+   -j - do restart on instances from the specified database table. If not specified, restart will apply on instances present in cloud.vm_instance database table. Works ony when restarting SSVMs, CPVMs or VRs\n\n"
```
Comment (weizhouapache) [inline setup/bindir/cloud-sysvmadm.in:44]: is there any table than vm_instance ?

File: setup/bindir/cloud-sysvmadm.in (hunk 3 above)
Comment (shwstppr) [reply setup/bindir/cloud-sysvmadm.in]: :facepalm: fixed

File: setup/bindir/cloud-sysvmadm.in (hunk 4 above)
Comment (shwstppr) [reply setup/bindir/cloud-sysvmadm.in]: @weizhouapache fixed

File: setup/bindir/cloud-sysvmadm.in (hunk 5 above)
Comment (shwstppr) [reply setup/bindir/cloud-sysvmadm.in]: @weizhouapache the original issue requested option to provide a different VM table,

> Parametrize table name allowing cloning subsets to temporary data tables in the DB

Maybe the use-case is to copy a subset of entries in a new table and use the script on those VM entries

Comment (weizhouapache) [reply setup/bindir/cloud-sysvmadm.in]: @shwstppr 
ok, make sense

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by shwstppr** [comment]
@DaanHoogland I don't think the script, this PR changes, is called by the management server or any agent

**Pull request merged**